src/Package.zig

```diff
@@ -130,15 +118,97 @@ pub fn deinitTable(pkg: *Package, gpa: Allocator) void {
     pkg.table.deinit(gpa);
 }
 
-pub fn add(pkg: *Package, gpa: Allocator, package: *Package) !void {
+pub fn add(pkg: *Package, gpa: Allocator, name: []const u8, package: *Package) !void {
     try pkg.table.ensureUnusedCapacity(gpa, 1);
-    pkg.table.putAssumeCapacityNoClobber(package.name, package);
+    const name_dupe = try gpa.dupe(u8, name);
+    pkg.table.putAssumeCapacityNoClobber(name_dupe, package);
 }
 
-pub fn addAndAdopt(parent: *Package, gpa: Allocator, child: *Package) !void {
-    assert(child.parent == null); // make up your mind, who is the parent??
-    child.parent = parent;
-    return parent.add(gpa, child);
+/// Compute a readable name for the package. The returned name should be freed from gpa. This
+/// function is very slow, as it traverses the whole package hierarchy to find a path to this
+/// package. It should only be used for error output.
+pub fn getName(target: *const Package, gpa: Allocator, mod: Module) ![]const u8 {
```

Comment:
This would be the first breadth-first search added to the compiler codebase. Instead, could this name be constructed during CLI parsing while resolving module dependencies?

Reply:
The point of the search is that we want to find a relatively *short* name for the module where possible. You can imagine some library `foo` which you depend on directly but is also really deep in your dependency hierarchy, so that it's both legal to reference it as `root.foo` and `root.thing.bar.http.dep.foo` - in this case it's much more enlightening to give the first name, hence the breadth-first search. Unless I'm missing something (which is very possible, I just woke up) I can't see a way we could construct that during CLI parsing without effectively, well, doing a breadth-first search.

Reply:
I mistakenly thought that the CLI was recursively iterating over all the modules, so that idea itself wouldn't work. But perhaps there is still merit to it. Idea being that a single iteration up front over all the modules could establish a "shortest parent" for each module, which could then be used later in order to construct the lookup path without a search. I think the autodoc system wants this information too.

The way you have it now is decently encapsulated, so I think it's OK, but I have a hunch that computing the info up front will be generally useful for other things besides error reporting as well.

Reply:
Okay, that sounds reasonable. I'm not entirely sure about whether this would be a good idea for anything other than error reporting in its current state, though: hashmap iteration order means its output can be inconsistent if there are multiple same-length paths to the module. A system for use anywhere else should probably do something to make it consistent (where does autodocs want this data out of interest?). Do you want me to try and change it now, or should I leave it to a future PR?

Reply:
I think you can leave it for now. You've solved a major problem, and this other (admittedly ill-defined) problem can and should be tackled separately.